src/h3/_cy/memory.pyx

```diff
@@ -0,0 +1,210 @@
+        size_t j = n
+
+    while i < j:
+        if a[j-1] == 0:
+            j -= 1
+            continue
+
+        if a[i] != 0:
+            i += 1
+            continue
+
+        # if we're here, we know:
+        # a[i] == 0
+        # a[j-1] != 0
+        # i < j
+        # so we can swap! (actually, move a[j-1] -> a[i])
+        a[i] = a[j-1]
+        j -= 1
+
+    return i
+
+
+cdef H3int[:] empty_memory_view():
+    # todo: get rid of this?
+    # there's gotta be a better way to do this...
+    # create an empty cython.view.array?
+    cdef:
+        H3int a[1]
+
+    return (<H3int[:]>a)[:0]
+
+
+cdef _remove_zeros(H3MemoryManager x):
+    x.n = move_nonzeros(x.ptr, x.n)
+    if x.n == 0:
+        h3_free(x.ptr)
+        x.ptr = NULL
+
+    x.ptr = <H3int*> h3_realloc(x.ptr, x.n*sizeof(H3int))
+    if not x.ptr:
+        raise MemoryError()
+
+
+cdef H3int[:] _create_mv(H3MemoryManager x):
+    cdef:
+        array mv
+
+    if x.n == 0:
+        return empty_memory_view()
+
+    mv = <H3int[:x.n]> x.ptr
+    mv.callback_free_data = h3_free
+
+    # responsibility for the memory moves from this object to the array/memoryview
+    x.ptr = NULL
+    x.n = 0
+
+    return mv
+
+
+cdef class H3MemoryManager:
+    def __cinit__(self, size_t n):
+        self.n = n
+        self.ptr = <H3int*> h3_calloc(self.n, sizeof(H3int))
+        # todo: do we actually have a memory leak when n = 0 because of the non-null pointer thing?
+
+        if not self.ptr:
+            raise MemoryError()
+
+    cdef H3int[:] to_mv(self):
+        _remove_zeros(self)
+        return _create_mv(self)
+
+    def __dealloc__(self):
+        # If the memory has been handed off to a memoryview, this pointer
+        # should be NULL, and deallocing on NULL is fine.
+        # If the pointer is *not* NULL, then this means the MemoryManager
+        # has is still responsible for the memory (it hasn't given the memory away to another object).
+        h3_free(self.ptr)
+
+
+"""
+Can someone please swoop in and find a much cleaner way to do all of this Cython memory management baloney?
+"""
+cdef int[:] int_mv(size_t n):
+    cdef:
+        int* ptr
+        array arr
+
+    if n <= 0:
+        raise MemoryError()
+    ptr = <int*> h3_calloc(n, sizeof(int))
+    if ptr is NULL:
+        raise MemoryError()
+
+    arr = <int[:n]> ptr
+    arr.callback_free_data = h3_free
+
+    return arr
+
+
+cdef H3int[:] simple_mv(size_t n):
+    # cdef:
+    #     array mv
+
+    if n == 0:
+        return empty_memory_view()
+
+    ptr = <H3int*> h3_calloc(n, sizeof(H3int))
+    if not ptr:
+        raise MemoryError()
+
+    mv = <H3int[:n]> ptr
+    mv.callback_free_data = h3_free
+
+    return mv
+
+
+cpdef H3int[:] iter_to_mv(hexes):
+    """ hexes needs to be an iterable that knows its size...
+    or should we have it match the np.fromiter function, which infers if not available?
+    """
+    # cdef array x  # this needs to be commented out to avoid an error
```

Comment:
delete unused line? is the comment semantically relevant?

Reply:
Figured out what was going on. removed.